logstash-core/src/main/java/org/logstash/execution/AbstractPipelineExt.java

```diff
@@ -163,10 +164,24 @@ public class AbstractPipelineExt extends RubyBasicObject {
     private @SuppressWarnings("rawtypes") RubyArray filters;
     private @SuppressWarnings("rawtypes") RubyArray outputs;
 
+    private String lastErrorEvaluationReceived = "";
+
     public AbstractPipelineExt(final Ruby runtime, final RubyClass metaClass) {
         super(runtime, metaClass);
     }
 
+    public interface ConditionalEvaluationListener {
+        void notify(ConditionalEvaluationError err);
+    }
+
+    public final class LogErrorEvaluationListener implements ConditionalEvaluationListener {
+        @Override
+        public void notify(ConditionalEvaluationError err) {
+            lastErrorEvaluationReceived = err.getMessage();
+            LOGGER.warn("Error in condition evaluation with event {}", err.failedEvent().getField("[event][original]"), err);
```

Comment:
the issue with doing `LOGGER.warn("string {}", event, err)` is that it prints the exception on a new line:

```
[2024-07-24T12:42:34,289][WARN ][org.logstash.execution.AbstractPipelineExt][main] Error in condition evaluation with event null
org.logstash.config.ir.compiler.ConditionalEvaluationError: org.logstash.config.ir.compiler.EventCondition$Compiler$UnexpectedTypeException: Unexpected input type combination left <no-class>:<null-value>, right class org.jruby.RubyFixnum:100
	at org.logstash.config.ir.compiler.Utils.filterEvents(Utils.java:56) ~[logstash-core.jar:?]
	at org.logstash.generated.CompiledDataset1.compute(Unknown Source) ~[?:?]
	at org.logstash.generated.CompiledDataset2.compute(Unknown Source) ~[?:?]
	at org.logstash.generated.CompiledDataset3.compute(Unknown Source) ~[?:?]
	at org.logstash.config.ir.CompiledPipeline$CompiledUnorderedExecution.compute(CompiledPipeline.java:364) ~[logstash-core.jar:?]
	at org.logstash.config.ir.CompiledPipeline$CompiledUnorderedExecution.compute(CompiledPipeline.java:358) ~[logstash-core.jar:?]
	at org.logstash.execution.ObservedExecution.lambda$compute$0(ObservedExecution.java:17) ~[logstash-core.jar:?]
	at org.logstash.execution.WorkerObserver.lambda$observeExecutionComputation$0(WorkerObserver.java:39) ~[logstash-core.jar:?]
	at org.logstash.instrument.metrics.timer.ConcurrentLiveTimerMetric.time(ConcurrentLiveTimerMetric.java:47) ~[logstash-core.jar:?]
	at org.logstash.execution.WorkerObserver.lambda$executeWithTimers$1(WorkerObserver.java:50) ~[logstash-core.jar:?]
	at org.logstash.instrument.metrics.timer.ConcurrentLiveTimerMetric.time(ConcurrentLiveTimerMetric.java:47) ~[logstash-core.jar:?]
	at org.logstash.execution.WorkerObserver.executeWithTimers(WorkerObserver.java:50) ~[logstash-core.jar:?]
	at org.logstash.execution.WorkerObserver.observeExecutionComputation(WorkerObserver.java:38) ~[logstash-core.jar:?]
	at org.logstash.execution.ObservedExecution.compute(ObservedExecution.java:17) ~[logstash-core.jar:?]
	at org.logstash.execution.WorkerLoop.abortableCompute(WorkerLoop.java:113) ~[logstash-core.jar:?]
	at org.logstash.execution.WorkerLoop.run(WorkerLoop.java:86) ~[logstash-core.jar:?]
	at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:580) ~[?:?]
	at org.jruby.javasupport.JavaMethod.invokeDirectWithExceptionHandling(JavaMethod.java:300) ~[jruby.jar:?]
	at org.jruby.javasupport.JavaMethod.invokeDirect(JavaMethod.java:164) ~[jruby.jar:?]
	at org.jruby.java.invokers.InstanceMethodInvoker.call(InstanceMethodInvoker.java:32) ~[jruby.jar:?]
	at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:193) ~[jruby.jar:?]
	at org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:346) ~[jruby.jar:?]
	at org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(StartupInterpreterEngine.java:66) ~[jruby.jar:?]
	at org.jruby.ir.interpreter.Interpreter.INTERPRET_BLOCK(Interpreter.java:118) ~[jruby.jar:?]
	at org.jruby.runtime.MixedModeIRBlockBody.commonYieldPath(MixedModeIRBlockBody.java:136) ~[jruby.jar:?]
	at org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:66) ~[jruby.jar:?]
	at org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:58) ~[jruby.jar:?]
	at org.jruby.runtime.Block.call(Block.java:144) ~[jruby.jar:?]
	at org.jruby.RubyProc.call(RubyProc.java:354) ~[jruby.jar:?]
	at org.jruby.internal.runtime.RubyRunnable.run(RubyRunnable.java:111) ~[jruby.jar:?]
	at java.lang.Thread.run(Thread.java:1583) ~[?:?]
Caused by: org.logstash.config.ir.compiler.EventCondition$Compiler$UnexpectedTypeException: Unexpected input type combination left <no-class>:<null-value>, right class org.jruby.RubyFixnum:100
	at org.logstash.config.ir.compiler.EventCondition$Compiler.compare(EventCondition.java:458) ~[logstash-core.jar:?]
	at org.logstash.config.ir.compiler.EventCondition$Compiler.lambda$compareFieldToConstant$11(EventCondition.java:449) ~[logstash-core.jar:?]
	at org.logstash.config.ir.compiler.Utils.filterEvents(Utils.java:52) ~[logstash-core.jar:?]
	... 31 more
```

But when using log.format=json the exception doesn't show up:

```
{"level":"WARN","loggerName":"org.logstash.execution.AbstractPipelineExt","timeMillis":1721821446156,"thread":"[main]>worker8","logEvent":{"message":"Error in condition evaluation with event null"}}
```

Maybe we can log only the lastErrorEvaluationReceived at warn level, but add the rest to a debug log level entry.

Reply:
So we have this problem of loosing the exceptions for each exception logged if the format is JSON.
I think we could do something like:
```java
LOGGER.warn("Error in condition evaluation with error {} on event {}", lastErrorEvaluationReceived, err.failedEvent().getField("[event][original]"));
try (StringWriter sw = new StringWriter(); PrintWriter pw = new PrintWriter(sw)) {
    err.printStackTrace(pw);
    LOGGER.debug("{}", sw);
}
```

Which would generate the following JSON log:
```json
{"level":"DEBUG",
"loggerName":"org.logstash.execution.AbstractPipelineExt",
"timeMillis":1721835389378,
"thread":"[main]>worker1",
"logEvent":{
    "message":"org.logstash.config.ir.compiler.ConditionalEvaluationError: org.logstash.config.ir.compiler.EventCondition$Compiler$UnexpectedTypeException: Unexpected input type combination left class org.logstash.ConvertedList:ConvertedList{delegate=[101, 102]}, right class org.jruby.RubyFixnum:100\n\tat org.logstash.config.ir.compiler.Utils.filterEvents(Utils.java:56)\n\tat org.logstash.generated.CompiledDataset1.compute(Unknown Source)\n\tat org.logstash.generated.CompiledDataset2.compute(Unknown Source)\n\tat org.logstash.generated.CompiledDataset3.compute(Unknown Source)\n\tat org.logstash.config.ir.CompiledPipeline$CompiledUnorderedExecution.compute(CompiledPipeline.java:364)\n\tat org.logstash.config.ir.CompiledPipeline$CompiledUnorderedExecution.compute(CompiledPipeline.java:358)\n\tat org.logstash.execution.ObservedExecution.lambda$compute$0(ObservedExecution.java:17)\n\tat org.logstash.execution.WorkerObserver.lambda$observeExecutionComputation$0(WorkerObserver.java:39)\n\tat org.logstash.instrument.metrics.timer.ConcurrentLiveTimerMetric.time(ConcurrentLiveTimerMetric.java:47)\n\tat org.logstash.execution.WorkerObserver.lambda$executeWithTimers$1(WorkerObserver.java:50)\n\tat org.logstash.instrument.metrics.timer.ConcurrentLiveTimerMetric.time(ConcurrentLiveTimerMetric.java:47)\n\tat org.logstash.execution.WorkerObserver.executeWithTimers(WorkerObserver.java:50)\n\tat org.logstash.execution.WorkerObserver.observeExecutionComputation(WorkerObserver.java:38)\n\tat org.logstash.execution.ObservedExecution.compute(ObservedExecution.java:17)\n\tat org.logstash.execution.WorkerLoop.abortableCompute(WorkerLoop.java:113)\n\tat org.logstash.execution.WorkerLoop.run(WorkerLoop.java:86)\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)\n\tat java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\n\tat java.base/java.lang.reflect.Method.invoke(Method.java:568)\n\tat org.jruby.javasupport.JavaMethod.invokeDirectWithExceptionHandling(JavaMethod.java:300)\n\tat org.jruby.javasupport.JavaMethod.invokeDirect(JavaMethod.java:164)\n\tat org.jruby.java.invokers.InstanceMethodInvoker.call(InstanceMethodInvoker.java:32)\n\tat org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:456)\n\tat org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:195)\n\tat org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:346)\n\tat org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(StartupInterpreterEngine.java:66)\n\tat org.jruby.ir.interpreter.Interpreter.INTERPRET_BLOCK(Interpreter.java:118)\n\tat org.jruby.runtime.MixedModeIRBlockBody.commonYieldPath(MixedModeIRBlockBody.java:136)\n\tat org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:66)\n\tat org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:58)\n\tat org.jruby.runtime.Block.call(Block.java:144)\n\tat org.jruby.RubyProc.call(RubyProc.java:354)\n\tat org.jruby.internal.runtime.RubyRunnable.run(RubyRunnable.java:111)\n\tat java.base/java.lang.Thread.run(Thread.java:840)\nCaused by: org.logstash.config.ir.compiler.EventCondition$Compiler$UnexpectedTypeException: Unexpected input type combination left class org.logstash.ConvertedList:ConvertedList{delegate=[101, 102]}, right class org.jruby.RubyFixnum:100\n\tat org.logstash.config.ir.compiler.EventCondition$Compiler.compare(EventCondition.java:458)\n\tat org.logstash.config.ir.compiler.EventCondition$Compiler.lambda$compareFieldToConstant$11(EventCondition.java:449)\n\tat org.logstash.config.ir.compiler.Utils.filterEvents(Utils.java:52)\n\t... 34 more\n"
  }
}
```

Reply:
Fixed by ee1837adf6d3fca2cc3367ce4826b06031a85975